balloc/ut/balloc.c

```diff
@@ -263,52 +252,130 @@ int test_balloc_ut_ops(struct m0_be_ut_backend *ut_be, struct m0_be_seg *seg,
 
 out:
 	m0_free(prev_group_info_free_blocks);
-
 	M0_LOG(M0_INFO, "done. status = %d", rc);
 	return rc;
 }
 
-void test_balloc()
+void prepare_balloc_config(struct m0_ad_balloc_format_req *bcfg,
+			   m0_bcount_t totalsize,
+			   m0_bcount_t groupsize,
+			   m0_bcount_t indexcount)
 {
-	struct m0_be_ut_backend	 ut_be;
-	struct m0_be_ut_seg	 ut_seg;
-	int			 rc;
+	m0_bcount_t grp_count;
+
+	/*
+ 	 * group size is counted depending on BALLOC_DEF_GROUPS_NR.
+ 	 * Group size must be power of 2.
+ 	 */ 
+	groupsize = 1 << m0_log2(groupsize);
+	groupsize = max64u(groupsize, BALLOC_DEF_BLOCKS_PER_GROUP);
+	grp_count = m0_balloc_group_count(totalsize,
+					   (1 << BALLOC_DEF_BLOCK_SHIFT),
+					   groupsize);
+
+	bcfg->bfr_fid                   = M0_FID_INIT(0, 1);
+	bcfg->bfr_totalsize             = totalsize;
+	bcfg->bfr_blocksize             = (1 << BALLOC_DEF_BLOCK_SHIFT);
+	bcfg->bfr_groupsize             = groupsize;
+	bcfg->bfr_groupcount            = grp_count;
+	bcfg->bfr_indexcount            = indexcount;
+	bcfg->bfr_spare_reserved_blocks = m0_stob_ad_spares_calc(
+						groupsize);
+}
 
+void test_balloc_helper(struct m0_ad_balloc_format_req  *bcfg, bool is_reserve)
+{
+	struct m0_be_ut_backend	        ut_be;
+	struct m0_be_ut_seg	        ut_seg;
+	int			        rc;
 	M0_SET0(&ut_be);
 	/* Init BE */
 	m0_be_ut_backend_init(&ut_be);
 	m0_be_ut_seg_init(&ut_seg, &ut_be, 1ULL << 24);
-	rc = test_balloc_ut_ops(&ut_be, ut_seg.bus_seg, false);
+	rc = test_balloc_ut_ops(&ut_be, ut_seg.bus_seg, bcfg, is_reserve);
 	M0_UT_ASSERT(rc == 0);
 
 	m0_be_ut_seg_fini(&ut_seg);
 	m0_be_ut_backend_fini(&ut_be);
 }
 
+void test_balloc()
+{
+
+	struct m0_ad_balloc_format_req  bcfg;
+
+	prepare_balloc_config(&bcfg, BALLOC_DEF_CONTAINER_SIZE,
+			      BALLOC_DEF_BLOCKS_PER_GROUP,
+			      BALLOC_DEF_INDEXES_NR);
+	test_balloc_helper(&bcfg, false);	
+}
+
 void test_reserve_extent()
 {
-	struct m0_be_ut_backend	 ut_be;
-	struct m0_be_ut_seg	 ut_seg;
-	int			 rc;
+	struct m0_ad_balloc_format_req  bcfg;
 
-	M0_SET0(&ut_be);
-	/* Init BE */
-	m0_be_ut_backend_init(&ut_be);
-	m0_be_ut_seg_init(&ut_seg, &ut_be, 1ULL << 24);
-	rc = test_balloc_ut_ops(&ut_be, ut_seg.bus_seg, true);
-	M0_UT_ASSERT(rc == 0);
+	prepare_balloc_config(&bcfg, BALLOC_DEF_CONTAINER_SIZE,
+			      BALLOC_DEF_BLOCKS_PER_GROUP,
+			      BALLOC_DEF_INDEXES_NR);
+	test_balloc_helper(&bcfg, true);	
+}
 
-	m0_be_ut_seg_fini(&ut_seg);
-	m0_be_ut_backend_fini(&ut_be);
+void test_balloc_group_size()
+{
+	struct m0_ad_balloc_format_req  bcfg;
+	int                             iter;
+
+	/* Test various group size */
+	for (iter = 0; iter <= 4; iter++) {
+		prepare_balloc_config(&bcfg, BALLOC_DEF_CONTAINER_SIZE,
+				      BALLOC_DEF_BLOCKS_PER_GROUP * (1 << iter),
```

Comment:
In /etc/sysconfig/motr we configure MOTR_M0D_BALLOC_GROUP_COUNT where as in this test we are scaling BALLOC_DEF_BLOCKS_PER_GROUP which will also have similar impact, but would you line to update /etc/sysconfig/motr  also with MOTR_M0D_BALLOC_DEF_BLOCKS_PER_GROUP so that the configuration is in sync.

Reply:
Added new TC  for different group count including TC for group size.  did not disturb prepare_balloc_config() to include group count as parameter as  other TC will be affected..

Reply:
ok.